lib/events/athena/consumer.go

```diff
@@ -0,0 +1,457 @@
+	maxNumberOfMessagesFromReceive = 10
+)
+
+// consumer is responsible for receiving messages from SQS, batching them up to
+// certain size or interval, and writes to s3 as parquet file.
+type consumer struct {
+	logger              *log.Entry
+	backend             backend.Backend
+	storeLocationPrefix string
+	storeLocationBucket string
+	batchMaxItems       int
+	batchMaxInterval    time.Duration
+
+	collectConfig sqsCollectConfig
+}
+
+type sqsCollectConfig struct {
+	sqsReceiver       sqsReceiver
+	queueURL          string
+	payloadBucket     string
+	payloadDownloader s3downloader
+	// visibilityTimeout defines how long message won't be available for other
+	// receiveMessage calls. If timeout happens, and message was not deleted
+	// it will return to the queue.
+	visibilityTimeout     int32
+	waitOnReceiveTimeout  int32
+	waitOnReceiveDuration time.Duration
+
+	batchMaxItems int
+}
+
+type sqsReceiver interface {
+	ReceiveMessage(ctx context.Context, params *sqs.ReceiveMessageInput, optFns ...func(*sqs.Options)) (*sqs.ReceiveMessageOutput, error)
+}
+
+type s3downloader interface {
+	Download(ctx context.Context, w io.WriterAt, input *s3.GetObjectInput, options ...func(*manager.Downloader)) (n int64, err error)
+}
+
+func newConsumer(cfg Config, awsCfg aws.Config, backend backend.Backend, logEntry *log.Entry) (*consumer, error) {
+	// TODO(tobiaszheller): move this to Config.CheckAndSetDefaults.
+	u, err := url.Parse(cfg.LocationS3)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	largeEventsURL, err := url.Parse(cfg.LargeEventsS3)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	s3client := s3.NewFromConfig(awsCfg)
+	sqsReceiver := sqs.NewFromConfig(awsCfg)
+	return &consumer{
+		logger:              logEntry,
+		backend:             backend,
+		storeLocationPrefix: strings.TrimSuffix(strings.TrimPrefix(u.Path, "/"), "/"),
+		storeLocationBucket: u.Host,
+		batchMaxItems:       cfg.BatchMaxItems,
+		batchMaxInterval:    cfg.BatchMaxInterval,
+		collectConfig: sqsCollectConfig{
+			sqsReceiver: sqsReceiver,
+			queueURL:    cfg.QueueURL,
+			// TODO(tobiaszheller): use s3 manager from teleport observability.
+			payloadDownloader:     manager.NewDownloader(s3client),
+			payloadBucket:         largeEventsURL.Host,
+			visibilityTimeout:     int32(cfg.BatchMaxInterval.Seconds()),
+			waitOnReceiveTimeout:  int32(maxWaitTimeOnReceiveMessageFromSQS.Seconds()),
+			waitOnReceiveDuration: maxWaitTimeOnReceiveMessageFromSQS,
+			batchMaxItems:         cfg.BatchMaxItems,
+		},
+	}, nil
+}
+
+// run continuously runs batching job. It is blocking operation.
+// It is stopped via canceling context.
+func (c *consumer) run(ctx context.Context) {
+	const lockName = "athena_batcher"
+	// lockTTL is time after which other auth server can acquire lock.
+	// It can be half of BatchMaxInterval because RunWhileLocked automatically
+	// refresh TTL if job is still running.
+	lockTTL := c.batchMaxInterval / 2
+	var err error
+	for {
```

Comment:
Won't this loop cause a lot of pressure on auth when the cluster is idle?

When the cluster is idle, SQS will be empty, and this loop will try and write to auth (DynamoDB) every 250 ms.

What about using a channel (or a `sync.Cond`) to signal that there is data to process and an attempt should be made to lock auth?

Reply:
>Won't this loop cause a lot of pressure on auth when the cluster is idle?

It will, I left it as todo here to increase that timeout of 250ms. https://github.com/gravitational/teleport/blob/7d424230502658d0b5d4ec9e2c4aa075d66bd8f2/lib/events/athena/consumer.go#L124-L125

> What about using a channel (or a sync.Cond) to signal that there is data to process and an attempt should be made to lock auth?

We can use `sqs.GetQueueAttributes` to check how many messages are on queue ([ApproximateNumberOfMessages](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_GetQueueAttributes.html)).

However there still be pressure on backend during processing, 1 auth will process batch for 1 minute, and other will try every 250ms to acquire lock via dynamo. I think this timeout should be configured to 10s, WDYT?

Reply:
@russjones https://github.com/gravitational/teleport/pull/24559 here is PR with confiogurable retry timeout